### ISSUE TITLE
[JENKINS-27329] WorkspaceCleanupThread may delete workspaces of running jobs

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/jenkins#20929]
<p>The problem is as described in <a href="https://issues.jenkins.io/browse/JENKINS-4501" title="WorkspaceCleanupThread deletes Matrix project workspaces" class="issue-link" data-issue-key="JENKINS-4501"><del>JENKINS-4501</del></a>. As requested in <a href="https://issues.jenkins.io/browse/JENKINS-4501" title="WorkspaceCleanupThread deletes Matrix project workspaces" class="issue-link" data-issue-key="JENKINS-4501"><del>JENKINS-4501</del></a>, I am creating a new issue as this problem still exists in 1.602.</p>

<p>In short, Jenkins silently and erroneously deletes workspaces on slaves for matrix projects that are not old.</p>

<p>Over the course of the time I've worked with Jenkins this behavior has created literally days of work and waiting on very long running builds that rely on cached workspaces to be manageable. It's cost me more hours again today after restoring jenkins to a new server after a hardware failure. This setting was reset since it exists outside normal recommended backup files and I didn't think to add it when I "fixed" this last time.</p>

<p>Would it not be easier to have hudson.model.WorkspaceCleanupThread.disabled default to true? Having the default behavior be "destroy my data" seems bad, especially with how cheap disk is now. I'm sure when this option was implemented it made a lot of sense, but when I can get a 1TB for $50, it just seems wrong-headed. Let the fallow workspaces lie. I can clean them up if I need to.</p>

<p>If that's not an acceptable solution, could it not be moved to a config location in the Jenkins home? That way we can be relatively sure that the setting will be propagated in backups and not bite someone who thought they solved this problem and had forgotten about it?</p>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/qhartman.jpg" title="qhartman's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=qhartman">qhartman</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-27329" target="_blank">WorkspaceCleanupThread may delete workspaces of running jobs</a></i></summary>
<i><ul>
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>component(s)</b>: core
<li><b>label(s)</b>: 2.121.2-fixed
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2018-06-06T09:43:24+00:00
<li><b>votes</b>: 13
<li><b>watchers</b>: 19
<li><b>imported</b>: 2025-11-24
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>The problem is as described in <a href="https://issues.jenkins.io/browse/JENKINS-4501" title="WorkspaceCleanupThread deletes Matrix project workspaces" class="issue-link" data-issue-key="JENKINS-4501"><del>JENKINS-4501</del></a>. As requested in <a href="https://issues.jenkins.io/browse/JENKINS-4501" title="WorkspaceCleanupThread deletes Matrix project workspaces" class="issue-link" data-issue-key="JENKINS-4501"><del>JENKINS-4501</del></a>, I am creating a new issue as this problem still exists in 1.602.</p>

<p>In short, Jenkins silently and erroneously deletes workspaces on slaves for matrix projects that are not old.</p>

<p>Over the course of the time I've worked with Jenkins this behavior has created literally days of work and waiting on very long running builds that rely on cached workspaces to be manageable. It's cost me more hours again today after restoring jenkins to a new server after a hardware failure. This setting was reset since it exists outside normal recommended backup files and I didn't think to add it when I "fixed" this last time.</p>

<p>Would it not be easier to have hudson.model.WorkspaceCleanupThread.disabled default to true? Having the default behavior be "destroy my data" seems bad, especially with how cheap disk is now. I'm sure when this option was implemented it made a lot of sense, but when I can get a 1TB for $50, it just seems wrong-headed. Let the fallow workspaces lie. I can clean them up if I need to.</p>

<p>If that's not an acceptable solution, could it not be moved to a config location in the Jenkins home? That way we can be relatively sure that the setting will be propagated in backups and not bite someone who thought they solved this problem and had forgotten about it?</p></pre>
</details>
</details>
<ul><li><i>environment</i>: <code>Linux host, Linux, OSX, Windows, slaves. Jenkins version 1.602.</code></li></ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>mwebber:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-27329?focusedId=234136&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-234136">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>&gt;&gt; Jenkins silently and erroneously deletes workspaces on slaves for matrix projects that are not old
We just hit this problem (or what appears to be this problem) as well.</p>

<p>An extract from <tt>$JENKINS_HOME/Workspace clean-up.log</tt>:</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>Deleting /Users/dlshudson/jenkins_slave/workspace/dials_distribute on dials-mac-mini
Deleting /scratch/jenkins_slave/workspace/dials_distribute on dials-ws133
Deleting /scratch/jenkins_slave/workspace/dials_distribute on dials-ws154
</pre>
</div></div>
<p><em>side-note</em>: it's a shame those log lines are not time-stamped
The 3 mentioned workspaces are from a matrix project, and all workspaces had been accessed recently.</p>

<p>Presumably there is a bug in the workspace cleanup code that means it does not handle matrix projects correctly.</p>

<p>Note that the job configuration specifies 4 slaves: 1 by label, and 3 by individual nodes. The workspaces that were deleted were those on the 3 slaves that were specified as individual notes, but the workspace on the slave that was specified by label was not deleted. Possibly a clue o the bug?</p>

<p>The workaround is to set <tt>hudson.model.WorkspaceCleanupThread.disabled=true</tt>.</p>
</pre>
</details>

</ul>
</details>

<p>>> Jenkins silently and erroneously deletes workspaces on slaves for matrix projects that are not old<br/>
We just hit this problem (or what appears to be this problem) as well.</p>

<p>An extract from <tt>$JENKINS_HOME/Workspace clean-up.log</tt>:</p>


```
Deleting /Users/dlshudson/jenkins_slave/workspace/dials_distribute on dials-mac-mini
Deleting /scratch/jenkins_slave/workspace/dials_distribute on dials-ws133
Deleting /scratch/jenkins_slave/workspace/dials_distribute on dials-ws154

```
<p><em>side-note</em>: it's a shame those log lines are not time-stamped<br/>
The 3 mentioned workspaces are from a matrix project, and all workspaces had been accessed recently.</p>

<p>Presumably there is a bug in the workspace cleanup code that means it does not handle matrix projects correctly.</p>

<p>Note that the job configuration specifies 4 slaves: 1 by label, and 3 by individual nodes. The workspaces that were deleted were those on the 3 slaves that were specified as individual notes, but the workspace on the slave that was specified by label was not deleted. Possibly a clue o the bug?</p>

<p>The workaround is to set <tt>hudson.model.WorkspaceCleanupThread.disabled=true</tt>.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>bonefish:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-27329?focusedId=238858&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-238858">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Since <a href="https://issues.jenkins.io/browse/JENKINS-30916" title="workspace being deleted mid-build on slave" class="issue-link" data-issue-key="JENKINS-30916"><del>JENKINS-30916</del></a> has been closed as a duplicate: Here the ticket description only says that workspaces that aren't old are deleted. In fact a workspace can even be deleted while a build using the workspace is in progress. The lines from the system log for such a case:</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>Okt 13, 2015 3:29:27 AM INFORMATION hudson.slaves.CommandLauncher launch
slave agent launched for BonefishMac-Ubuntu-12.04
Okt 13, 2015 3:31:15 AM INFORMATION hudson.model.AsyncPeriodicWork$1 run
Started Workspace clean-up
Okt 13, 2015 3:31:21 AM INFORMATION hudson.model.Run execute
Bar-Nightly/label=Ubuntu-12.04 #222 main build action completed: FAILURE
</pre>
</div></div>
</pre>
</details>

</ul>
</details>

<p>Since <a href="https://issues.jenkins.io/browse/JENKINS-30916" title="workspace being deleted mid-build on slave" class="issue-link" data-issue-key="JENKINS-30916"><del>JENKINS-30916</del></a> has been closed as a duplicate: Here the ticket description only says that workspaces that aren't old are deleted. In fact a workspace can even be deleted while a build using the workspace is in progress. The lines from the system log for such a case:</p>


```
Okt 13, 2015 3:29:27 AM INFORMATION hudson.slaves.CommandLauncher launch
slave agent launched for BonefishMac-Ubuntu-12.04
Okt 13, 2015 3:31:15 AM INFORMATION hudson.model.AsyncPeriodicWork$1 run
Started Workspace clean-up
Okt 13, 2015 3:31:21 AM INFORMATION hudson.model.Run execute
Bar-Nightly/label=Ubuntu-12.04 #222 main build action completed: FAILURE

```

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>mwebber:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-27329?focusedId=283281&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-283281">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=davida2009" class="user-hover" rel="davida2009">davida2009</a> It's passed as a Java system property when you start Jenkins. Something like:</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>java -Dhudson.model.WorkspaceCleanupThread.disabled=true -jar jenkins.war
</pre>
</div></div>
<p>See <a href="https://wiki.jenkins-ci.org/display/JENKINS/Features+controlled+by+system+properties" class="external-link" target="_blank" rel="nofollow noopener">https://wiki.jenkins-ci.org/display/JENKINS/Features+controlled+by+system+properties</a></p>

<p>Exactly how you change this on your system will depend on how you installed Jenkins, and what scripts you use to start it.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=davida2009" class="user-hover" rel="davida2009">davida2009</a> It's passed as a Java system property when you start Jenkins. Something like:</p>


```
java -Dhudson.model.WorkspaceCleanupThread.disabled=true -jar jenkins.war

```
<p>See <a href="https://wiki.jenkins-ci.org/display/JENKINS/Features+controlled+by+system+properties" class="external-link" target="_blank" rel="nofollow noopener">https://wiki.jenkins-ci.org/display/JENKINS/Features+controlled+by+system+properties</a></p>

<p>Exactly how you change this on your system will depend on how you installed Jenkins, and what scripts you use to start it.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>hushp1pt:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-27329?focusedId=293740&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-293740">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Thanks to all who wrote on this bug in 2017. I think the recent activity helped me find this bug when I searched for something to explain what was happening. This workaround does seem to work and I'm very grateful.  </p>

<p>Respectfully, I only wish I'd found it when I searched for the same, last year. </p>

<p> </p>
</pre>
</details>

</ul>
</details>

<p>Thanks to all who wrote on this bug in 2017. I think the recent activity helped me find this bug when I searched for something to explain what was happening. This workaround does seem to work and I'm very grateful.  </p>

<p>Respectfully, I only wish I'd found it when I searched for the same, last year. </p>

<p> </p>
